src/lib/optimizer/strategy/semi_join_reduction_rule.cpp

```diff
@@ -0,0 +1,146 @@
+
+    // As multi-predicate joins are expensive, we do not want to create semi join reductions that use them. Instead, we
+    // look at each predicate of the join independently. We can do this as a JoinNode's predicates are conjunctive.
+    // Disjunctive predicates are currently not supported and if they were, they would be stored in a single
+    // join_predicates entry.
+    for (const auto& join_predicate : join_node->join_predicates()) {
+      const auto predicate_expression = std::dynamic_pointer_cast<BinaryPredicateExpression>(join_predicate);
+      DebugAssert(predicate_expression, "Expected BinaryPredicateExpression");
+      if (predicate_expression->predicate_condition != PredicateCondition::Equals) {
+        continue;
+      }
+
+      // Since semi join reductions might be beneficial for both sides of the join, we create this helper lambda,
+      // which can deal with both sides.
+      const auto reduce_if_beneficial = [&](const auto side_of_join) {
+        auto reduced_node = join_node->input(side_of_join);
+        auto original_cardinality = estimator->estimate_cardinality(reduced_node);
+
+        auto reducer_node = join_node->input(opposite_side(side_of_join));
+        auto reducer_node_cardinality = estimator->estimate_cardinality(reducer_node);
+
+        const auto semi_join_reduction_node = JoinNode::make(JoinMode::Semi, predicate_expression);
+        semi_join_reduction_node->comment = "Semi Reduction";
+        lqp_insert_node(join_node, side_of_join, semi_join_reduction_node);
+        semi_join_reduction_node->set_right_input(reducer_node);
+
+        const auto reduced_cardinality = estimator->estimate_cardinality(semi_join_reduction_node);
+
+        semi_join_reduction_node->set_right_input(nullptr);
+        lqp_remove_node(semi_join_reduction_node);
+
+        if (original_cardinality == 0 || reduced_cardinality / original_cardinality > MINIMUM_SELECTIVITY) {
+          return;
+        }
+
+        // For `t1 JOIN t2 on t1.a = t2.a` where a semi join reduction is supposed to be added to t1, `t1.a = t2.a`
+        // is the predicate_expression. The values from t1 are supposed to be filtered by looking at t1.a, which is
+        // called the reducer_side_expression.
+        const auto& reducer_side_expression =
+            expression_evaluable_on_lqp(predicate_expression->left_operand(), *reducer_node)
+                ? predicate_expression->left_operand()
+                : predicate_expression->right_operand();
+        DebugAssert(!expression_evaluable_on_lqp(reducer_side_expression, *join_node->input(side_of_join)),
+                    "Expected filtered expression to be uniquely evaluable on one side of the join");
+
+        // Currently, the right input of the semi join reduction (i.e., the reducer node) is the opposite input of the
+        // join_node. By walking down that LQP, we might find a suitable reducer node where predicate_expression
```

Comment:
Don't you mean something like "find a suitable place for the reducer node" instead of "find a suitable reducer node"?

Reply:
The "reducer" is the node that serves as the right input to the semi join reduction.